File: tailbone.js
'use strict';
/*
 * Bi-Directional data binding with AppEngine and the channel api
 */

window.tailbone = (function(window, document, undefined) {

function POST(url, data, callback) {
  if (XMLHttpRequest) {
    r = new XMLHttpRequest();
    r.onreadystatechange = function() {
      if (r.readyState == 4) {
        callback(JSON.parse(r.responseText));
      }
    };
    r.open('POST', url, true);
    r.setRequestHeader('Content-Type', 'application/json');
    r.send(data);
  } else {
    throw Error('Browser does not support XMLHttpRequest. ' +
        'Try adding modernizer to polyfill.');
  }
}

function GET(url, callback) {
  if (XMLHttpRequest) {
    r = new XMLHttpRequest();
    r.onreadystatechange = function() {
      if (r.readyState == 4) {
        callback(JSON.parse(r.responseText));
      }
    };

    r.open('GET', url, true);
    r.setRequestHeader('Content-Type', 'application/json');
    r.send();
  } else {
    throw Error('Browser does not support XMLHttpRequest. ' +
        'Try adding modernizer to polyfill.');
  }
}

function DELETE(url, callback) {
  if (XMLHttpRequest) {
    r = new XMLHttpRequest();
    r.onreadystatechange = function() {
      if (r.readyState == 4) {
        callback(JSON.parse(r.responseText));
      }
    };
    r.open('DELETE', url, true);
    r.setRequestHeader('Content-Type', 'application/json');
    r.send();
  } else {
    throw Error('Browser does not support XMLHttpRequest. ' +
        'Try adding modernizer to polyfill.');
  }
}


/////////////////////////
// Events via Channel API
/////////////////////////

var CONNECTED = false;
var CONNECTING = false;
var BACKOFF = 1;

var event_map = {};
var queue = [];
var socket;
var client_id = parseInt(Date.now() / Math.random());

function onOpen() {
  CONNECTED = true;
  BACKOFF = 1;
  for (var i = 0, l = queue.length; i < l; i++) {
    queue[i].call(this);
  }
  queue = [];
}

function rebind() {
  var event_bindings = [];
  for (var name in event_map) {
    for (var fn in event_map[name]) {
      event_bindings.push([name, fn]);
    }
  }
  event_map = {};
  for (var i = 0, l = event_bindings.length; i < l; i++) {
    if_connected(bind, event_bindings[i][0], event_bindings[i][1]);
  }
}

function onClose() {
  CONNECTED = false;
  CONNECTING = false;
  rebind();
}

function onError() {
  // TODO: try reconnecting with backoff or alert system of lack of capability.
  throw new Error('Channel not connectable.');
}

function onMessage(msg) {
  var data = JSON.parse(msg.data);
  if (data.reboot) {
    // close everything and try and reconnect and rebind the event listeners
    socket.close();
    return;
  }
  var fns = event_map[data.name];
  if (fns && fns.length > 0) {
    for (var i = 0, l = fns.length; i < l; i++) {
      fns[i].call(this, data.payload);
    }
  }
}

function connect(callback) {
  if (callback) queue.push(callback);
  if (!CONNECTING) {
    post('/api/events/',
        JSON.stringify({'method': 'token', 'client_id': client_id}),
        function(resp) {
          var channel = new goog.appengine.Channel(resp.token);
          socket = channel.open();
          socket.onopen = onOpen;
          socket.onmessage = onMessage;
          socket.onclose = onClose;
          socket.onerror = onError;
        }
    );
  } else {
    CONNECTING = true;
  }
}

function if_connected(fn, arg1, arg2) {
  var _this = this;
  if (CONNECTED) {
    fn.call(_this, arg1, arg2);
  } else {
    connect(function() {fn.call(_this, arg1, arg2);});
  }
}

function errorHandler(msg) {
  if (msg && msg.error) {
    throw new Error(msg.error);
  }
}

function trigger(name, payload) {
  POST('/api/events/',
      JSON.stringify({'method': 'trigger',
                      'client_id': client_id,
                      'name': name, 'payload': payload}),
      errorHandler);
}

function bind(name, fn) {
  event_map[name] = event_map[name] || [];
  event_map[name].push(fn);
  POST('/api/events/',
      JSON.stringify({'method': 'bind',
                      'client_id': client_id, 'name': name}),
      errorHandler);
}

function unbind(name, fn) {
  if (name) {
    if (fn) {
      var index = event_map[name].indexOf[fn];
      delete event_map[name][index];
    } else {
      delete event_map[name];
    }
  } else {
    event_map = {};
  }
  POST('/api/events/',
      JSON.stringify({'method': 'unbind',
                      'client_id': client_id, 'name': name}),
      errorHandler);
}


/////////////////////////
// Quering and Filtering
/////////////////////////

function FILTER(name, opsymbol, value) {
  return [name, opsymbol, value];
}

function ORDER(name) {
  return name;
}

function AND() {
  var f = ['AND'];
  for (var i = 0; i < arguments.length; i++) {
    f.push(arguments[i]);
  }
  return f;
}

function OR() {
  var f = ['OR'];
  for (var i = 0; i < arguments.length; i++) {
    f.push(arguments[i]);
  }
  return f;
}

function save(type, model) {
}

function del(type, model) {
}


var ModelFactory = function(type, opt_schema) {
  ignored_prefixes = ['_', '$'];

  /**
  * Query is an iterable collection of a Model
  */
  var Query = function() {
    this.filter = [];
    this.order = [];
    this._page_size = 100;
    this._more = false;
    this._dirty = false;
  };

  Query.prototype = new Array();

  /**
  * onChange callback function
  */
  Query.prototype.onchange = undefined;

  Query.prototype.next = function() {
    return true;
  };

  Query.prototype.previous = function() {
    return true;
  };

  Query.prototype.__defineGetter__('more', function() { return this._more; });

  Query.prototype.__defineGetter__('page_size',
      function() { return this._page_size; });
  Query.prototype.__defineSetter__('page_size',
      function(page_size) {
        this._page_size = page_size;
        this._dirty = true;
      });

  Query.prototype.filter = function() {
    switch (arguments.length) {
      case 1:
        this.filter = this.filter.concat(arguments[0]);
        break;
      case 3:
        this.filter = this.filter.concat(FILTER.call(arguments));
        break;
      default:
        throw Error('Undefined FILTER format.');
    }
  };

  Query.prototype.order = function(name) {
    this.order.push(ORDER(name));
  };

  Query.prototype.to_json = function() {
    return {
      filter: this.filter,
      order: this.order,
      page_size: this._page_size
    };
  };

  Query.prototype.fetch = function(opt_callback) {
    var _this = this;
    function callback(data) {
      var fn = opt_callback || _this.onchange;
      if (fn) {
        fn(_this);
      }
    }
    GET('/api/' + type + '?params=' + JSON.stringify(this.to_json()), callback);
  };

  function update(model, data) {
    for (var k in data) {
      model[k] = data[k];
    }
  }

  var Model = function() {
  };

  /*
   * Get a model by its id.
   */
  Model.get = function(id, opt_callback) {
    var m = new Model();
    m.Id = id;
    function callback(data) {
      update(m, data);
      var fn = opt_callback || m.onchange;
      if (fn) {
        fn(m);
      }
    }
    GET('/api/' + type + '/' + id, callback);
    return m;
  };

  /*
   * query generates a iterator for a query object.
   */
  Model.query = function(opt_callback) {
    var query = new Query();
    // xhr query for collection with timeout to allow for chaining
    setTimeout(function() {
      query.fetch(opt_callback);
    }, 0);
    tailbone.bind(type, query.fetch);
    return query;
  };

  Model.to_json = function(model) {
    var obj = {};
    return obj;
  };

  /*
  * Dollar sign prefix names are ignored on the model as are underscore
  * _name
  * and
  * $name
  * are not included in the jsonifying of an object
  */
  Model.prototype.$save = function(opt_callback) {
    save(type, this);
    tailbone.trigger(type);
  };

  Model.prototype.$delete = function(opt_callback) {
    del(type, this);
    tailbone.trigger(type);
  };

  return Model;
};


// Add the channel js for appengine and bind the events.


// Exports
return {
  Model: ModelFactory,
  FILTER: FILTER,
  ORDER: ORDER,
  AND: AND,
  OR: OR,
  trigger: function(name, payload) { if_connected(trigger, name, payload); },
  bind: function(name, fn) { if_connected(bind, name, fn); },
  unbind: function(name, fn) { if_connected(unbind, name, fn); }
};

})(this, this.document);

File: tailbone.py
"""
tailbone

Appengine abstract restful backend
"""

import datetime
import functools
import json
import logging
import os
import random
import re
import sys
import time
import webapp2
import yaml

from google.appengine import api
from google.appengine.api import channel
from google.appengine.ext import ndb
from google.appengine.ext import deferred


"""
Intention:
  Try to stay as direct to the appengine api where applicable so it is an easy transition if you
  need to extend this.
  Make the api abstract enough so that any javascript framework will do and so that the same
  frontend could be used with another backend that implements the same api calls.

Ideas:
  Scoping private/public
    Capital is public
    lowercase is private

  Open Questions:
    cron job to look for data inconsistencies to notify you if they come up
    so you can ban a user if they are being a problem by storing things in the data set incorrectly.

    Send a weekly summary email about statistical outliers and information in your database.

    Special names
      CreatedAt or created_at -> auto sets the created at property and can only be modified by the
      server.
      ModifiedAt or modified_at -> where the server sets when it was last modified.

    How do you do full text search?
      Searchable{Text} -> if name is searchable it adds this to the full text search api
      /api/search/model?q=text

    Realtime binding
      Todo = new tailbone.Model("todos");
      todos = Todo.query({name: "doug"});
      t = new Todo({name: "doug");
      t.$save()
      t.$delete()

      channel listeners bound to queries
      save checks all listened queries for a matching one
"""

# Custom Exceptions
class AppError(Exception):
  pass

class BreakError(Exception):
  pass

class LoginError(Exception):
  pass


def current_user(required=False):
  u = api.users.get_current_user()
  if u:
    return u.user_id()
  if required:
    raise LoginError("User must be logged in.")
  return None

# Explicit Models
class ScopedExpando(ndb.Expando):
  owners__ = ndb.IntegerProperty(repeated=True)
  editors__ = ndb.IntegerProperty(repeated=True)
  def to_dict(self, *args, **kwargs):
    excluded = ["owners__","editors__"]
    if len(args) == 2:
      args[1] += exluded
    if kwargs.has_key("exclude"):
      kwargs["exclude"] += excluded
    else:
      kwargs["exclude"] = excluded
    result = super(ScopedExpando, self).to_dict(*args, **kwargs)
    # if the current user is owner return private attibutes too
    result["Id"] = self.key.id()
    return result

class users(ndb.Expando):
  def to_dict(self, *args, **kwargs):
    result = super(users, self).to_dict(*args, **kwargs)
    result["Id"] = self.key.id()
    return result

def json_extras(obj):
  """Extended json processing of types."""
  if hasattr(obj, "utctimetuple"):
    ms = time.mktime(obj.utctimetuple()) * 1000
    ms += getattr(obj, "microseconds", 0) / 1000
    return int(ms)
  return None

def as_json(func):
  """Returns json when callback in url"""
  @functools.wraps(func)
  def wrapper(self, *args, **kwargs):
    self.response.headers["Content-Type"] = "application/json"
    if DEBUG:
      self.response.headers["Access-Control-Allow-Origin"] = "*"
      self.response.headers["Access-Control-Allow-Methods"] = "POST,GET,PUT,PATCH,HEAD,OPTIONS"
      self.response.headers["Access-Control-Allow-Headers"] = "Content-Type"
    else:
      api.namespace_manager.set_namespace(VERSION)
    try:
      resp = func(self, *args, **kwargs)
    except BreakError as e:
      return
    except AppError as e:
      resp = { "error": str(e) }
      logging.error(str(e))
    except LoginError as e:
      url = users.create_login_url(self.request.url)
      resp = {
        "error": str(e),
        "url": url
      }
      logging.error(str(e))
    except api.datastore_errors.BadArgumentError as e:
      resp = { "error": str(e) }
      logging.error(str(e))
    if not isinstance(resp, str):
      resp = json.dumps(resp, default=json_extras)
    callback = self.request.get("callback")
    if callback:
      self.response.headers["Content-Type"] = "text/javascript"
      resp = "%s(%s);" % (_callback, resp)
    self.response.out.write(resp)
  return wrapper

def reflective_create(cls, data):
  m = cls()
  for k,v in data.iteritems():
    m._default_indexed = True
    if type(v) in [str, unicode]:
      if len(bytearray(v, encoding="utf8")) >= 500:
        m._default_indexed = False
    elif type(v) == dict:
      subcls = unicode.encode(k, "ascii", errors="ignore")
      v = reflective_create(type(subcls, (ndb.Expando,), {}), v)
    elif type(v) in [int, float]:
      v = float(v)
    setattr(m, k, v)
  return m

def parse_body(self):
  if "application/json" in self.request.content_type:
    data = json.loads(self.request.body)
  else:
    # TODO: must do further processing this into an dict
    # TODO: must upload any FieldStorage objects to blobstore
    data = self.request.params
  return data or {}

def clean_data(data):
  # strips any disallowed names {id, _*, etc}
  disallowed = ["Id", "id", "key"]
  exceptions = ["Id"]
  for key in disallowed:
    if data.has_key(key):
      if key not in exceptions:
        logging.warn("Disallowed key {%s} passed in object creation." % key)
      del data[key]
  return data


def parse_id(id, data_id=None):
  try:
    id = int(id)
  except:
    pass
  if data_id != None:
    if id:
      if data_id != id:
        raise AppError("Url id {%s} must match object id {%s}" % (id, data_id))
    else:
      id = data_id
  return id

re_filter = re.compile(r"^([\w\-.]+)(!=|==|=|<=|>=|<|>)(.+)$")
re_composite_filter = re.compile(r"^(AND|OR)\((.*)\)$")
re_split = re.compile(r",\W*")

def construct_filter(filter_str):
  m = re_composite_filter.match(filter_str)
  if m:
    filters = [construct_filter(f) for f in re_split.split(m.group(2))]
    if m.group(1) == "AND":
      return ndb.query.AND(*filters)
    else:
      return ndb.query.OR(*filters)
  m = re_filter.match(filter_str)
  if m:
    name, opsymbol, value = m.groups()
    if value == "true":
      value = True
    elif value == "false":
      value = False
    else:
      try:
        value = float(value)
      except:
        pass
    if opsymbol == "==":
      opsymbol = "="
    return ndb.query.FilterNode(name, opsymbol, value)
  if re_split.match(filter_str):
    return construct_filter("AND({})".format(filter_str))
  raise AppError("Filter format is unsupported: {}".format(filter_str))

def construct_order(cls, o):
  neg = True if o[0] == "-" else False
  o = o[1:] if neg else o
  if hasattr(cls, o):
    p = getattr(cls, o)
  else:
    p = ndb.GenericProperty(o)
  return -p if neg else p

def construct_query(cls, filters, orders):
  q = cls.query()
  q = q.filter(*[construct_filter(f) for f in filters])
  # TODO(doug) correctly auto append orders when necessary like on a multiselect
  q = q.order(*[construct_order(cls,o) for o in orders])
  return q

class RestfulHandler(webapp2.RequestHandler):
  @as_json
  def options(self, model, id):
    pass
  @as_json
  def get(self, model, id):
    # TODO(doug) does the model name need to be ascii encoded since types don't support utf-8
    cls = type(model.lower(), (ScopedExpando,), {})
    if id:
      id = parse_id(id)
      m = cls.get_by_id(id)
      if not m:
        raise AppError("There does not exists a %s with id %s" % (model, id))
      return m.to_dict()
    else:
      page_size = int(self.request.get("page_size", default_value=100))
      cursor = self.request.get("cursor")
      cursor = Cursor.from_websafe_string(cursor) if cursor else None
      projection = self.request.get("projection")
      filters = self.request.get_all("filter")
      orders = self.request.get_all("order")
      q = construct_query(cls, filters, orders)

      results, cursor, more = q.fetch_page(page_size=page_size, cursor=cursor)
      self.response.headers["More"] = "true" if more else "false"
      if cursor:
        self.response.headers["Next-Cursor"] = cursor.urlsafe()
        self.response.headers["Prev-Cursor"] = cursor.reversed().urlsafe()
      return [m.to_dict() for m in results]
  def set_or_create(self, model, id):
    cls = type(model.lower(), (ScopedExpando,), {})
    data = parse_body(self)
    id = parse_id(id, data.get("Id"))
    clean_data(data)
    m = reflective_create(cls, data)
    if id:
      m.key = ndb.Key(model, id)
    m.put()
    return m.to_dict()
  @as_json
  def post(self, model, id):
    return self.set_or_create(model, id)
  @as_json
  def patch(self, model, id):
    # TODO: implement this differently to do partial update
    return self.set_or_create(model, id)
  @as_json
  def put(self, model, id):
    return self.set_or_create(model, id)
  @as_json
  def delete(self, model, id):
    if not id:
      raise AppError("Must provide an id.")
    id = parse_id(id)
    key = ndb.Key(model.lower(), id)
    key.delete()
    return {}

class LoginHandler(webapp2.RequestHandler):
  def get(self):
    self.redirect(
        api.users.create_login_url(
          self.request.get("url", default_value="/")))

class LogoutHandler(webapp2.RequestHandler):
  def get(self):
    self.redirect(
        api.users.create_logout_url(
          self.request.get("url", default_value="/")))

class UsersHandler(webapp2.RequestHandler):
  """
  GET /api/users/:id
  returns the user info
  GET /api/users/me
  returns the current user
  """
  @as_json
  def get(self, id):
    pass
  @as_json
  def put(self, id):
    u = current_user(required=True)
    if id != "me" and id != u:
      raise AppError("Id must be the current user_id or me.")
    m = users.get_by_id(u)
    update_model(m, self)
    return m
  def delete(self, id):
    """Delete the user"s account and all their associated data."""
    pass

class AccessHandler(webapp2.RequestHandler):
  """
  GET /api/access/model/:id
    See if you have access to this model.
  {
    owners:[user_id],
    editors:[],
  }

  PUT /api/access/model/:id
    Give user with id user_id access
  {
    owners:[user_id],
    editors:[],
  }
  """
  @as_json
  def get(self, model, id):
    pass
  @as_json
  def put(self, model, id):
    pass

class AdminHandler(webapp2.RequestHandler):
  """Admin routes"""
  @as_json
  def get(self, action):
    if not api.users.is_current_user_admin():
      raise LoginError("You must be an admin.")


#-----------------
# START Event code
#-----------------
class events(ndb.Model):
  NUM_SHARDS = 20
  name = ndb.StringProperty()
  shard_id = ndb.IntegerProperty()
  listeners = ndb.IntegerProperty(repeated=True)


def bind(user_key, name):
  event = events.query(events.listeners == user_key, events.name == name).get()
  if not event:
    def txn():
      shard_id = random.randint(0, events.NUM_SHARDS - 1)
      event_key = ndb.Key(events, "{}_{}".format(name, shard_id))
      event = event_key.get()
      if not event:
        event = events(name=name, shard_id=shard_id, key=event_key)
      event.listeners.append(user_key)
      event.put()
      return event
    event = ndb.transaction(txn)
  return event

def unbind(user_key, name=None):
  eventlist = events.query(events.listeners == user_key)
  if name:
    eventlist = eventlist.filter(events.name == name)
  modified = []
  for event in eventlist:
    def txn():
      event.listeners = [l for l in event.listeners if l != user_key]
      if event.listeners:
        event.put()
      else:
        event.key.delete()
    ndb.transaction(txn)
    modified.append(event.to_dict())
  return modified

def trigger(name, payload):
  msg = json.dumps({ "name": name,
                     "payload": payload })
  sent_to = set()
  for e in events.query(events.name == name):
    for l in e.listeners:
      if l not in sent_to:
        sent_to.add(l)
        deferred.defer(channel.send_message, str(l), msg)
  return sent_to


class ConnectedHandler(webapp2.RequestHandler):
  @as_json
  def post(self):
    client_id = self.request.get('from')
    try:
      client_id = int(client_id)
    except:
      pass
    logging.info("Connecting client id {}".format(client_id))


class DisconnectedHandler(webapp2.RequestHandler):
  @as_json
  def post(self):
    client_id = self.request.get('from')
    try:
      client_id = int(client_id)
    except:
      pass
    logging.info("Disconnecting client id {}".format(client_id))
    unbind(client_id)


class RebootHandler(webapp2.RequestHandler):
  @as_json
  def get(self):
    # remove all event bindings and
    # force all current listeners to close and reconnect.
    logging.info("REBOOT")
    send_to = set()
    to_delete = set()
    msg = json.dumps({ "reboot": True })
    for e in events.query():
      for l in e.listeners:
        if l not in send_to:
          sent_to.add(l)
      to_delete.add(e.key)
    def delete():
      ndb.delete_multi(to_delete)
    ndb.transaction(delete)
    for l in send_to:
      deferred.defer(channel.send_message, str(l), msg)


class EventsHandler(webapp2.RequestHandler):
  @as_json
  def post(self):
    data = parse_body(self)
    method = data.get("method")
    client_id = data.get("client_id")
    if method == "token":
      return {"token": channel.create_channel(str(client_id))}
    elif method == "bind":
      bind(client_id, data.get("name"))
    elif method == "unbind":
      unbind(client_id, data.get("name"))
    elif method == "trigger":
      trigger(data.get("name"), data.get("payload"))

# ---------------
# END Event Code
# ---------------


APP_YAML = yaml.load(open("app.yaml"))
# prefix is taken from parsing the app.yaml
PREFIX = "/api/"

# VERSION is used to set the namespace
VERSION = APP_YAML.get("version")
DEBUG = os.environ.get("SERVER_SOFTWARE", "").startswith("Dev")

app = webapp2.WSGIApplication([
  ("/_ah/channel/connected/", ConnectedHandler),
  ("/_ah/channel/disconnected/", DisconnectedHandler),
  (r"{}login".format(PREFIX), LoginHandler),
  (r"{}logout" .format(PREFIX), LogoutHandler),
  (r"{}admin/(.+)".format(PREFIX), AdminHandler),
  (r"{}users/(.*)".format(PREFIX), UsersHandler),
  (r"{}access/([^/]+)/?(.*)".format(PREFIX), AccessHandler),
  (r"{}events/.*".format(PREFIX), EventsHandler),
  (r"{}([^/]+)/(.*)".format(PREFIX), RestfulHandler),
  ], debug=DEBUG)




